packages/resolver/lib/sources/fs.ts

```diff
@@ -23,38 +22,73 @@ export class FS implements ResolverSource {
       }
     }
 
-    try {
-      const result = fs.readFileSync(
-        path.join(searchPath, `${contractName}.json`),
-        "utf8"
-      );
-      return JSON.parse(result);
-    } catch (e) {
-      return null;
-    }
+    const result = this.searchForArtifact(normalizedImportPath, searchPath);
+    return result;
   }
 
-  getContractName(
+  searchForArtifact(
     sourcePath: string,
     searchPath = this.contractsBuildDirectory
   ) {
+    // most cases we can use the `{name}.sol -> {name}.json` convention
+    // console.log(sourcePath);
+    // console.log(searchPath);
+    const likelyFileBasename = path.basename(sourcePath, ".sol");
+    // console.log(likelyFileBasename);
+    // console.log(path.join(searchPath, `${likelyFileBasename}.json`));
+    const likelyArtifactString = fs.readFileSync(
+      path.join(searchPath, `${likelyFileBasename}.json`),
+      "utf8"
+    );
+
+    let likelyArtifact;
+    try {
+      if (likelyArtifactString) {
+        likelyArtifact = JSON.parse(likelyArtifactString);
+        return likelyArtifact;
+      }
+    } catch (e) {
+      // do nothing, we'll handle this by doing a deeper search
+      // console.log(e);
+    }
+
+    // we couldn't use the `{name}.sol --> {name}.json` convention, let's search
+
     const contractsBuildDirFiles = fs.readdirSync(searchPath);
+    // console.log(contractsBuildDirFiles);
     const filteredBuildArtifacts = contractsBuildDirFiles.filter(
-      (file: string) => file.match(".json") != null
+      // we don't care about files that don't have the json extension
+      // we also don't care about the file we already tried
+      (file: string) =>
+        file.match(".json") != null && file !== `${likelyFileBasename}.json`
     );
 
     for (const buildArtifact of filteredBuildArtifacts) {
-      const artifact = JSON.parse(
-        fs.readFileSync(path.resolve(searchPath, buildArtifact)).toString()
+      const artifactString = fs.readFileSync(
+        path.resolve(searchPath, buildArtifact),
+        { encoding: "utf8" }
       );
+      try {
+        if (artifactString) {
+          const artifact = JSON.parse(artifactString);
+
+          // console.log(artifact.sourcePath, sourcePath);
 
-      if (artifact.sourcePath === sourcePath) {
-        return artifact.contractName;
+          if (
+            path.basename(artifact.sourcePath, ".sol") ===
+              path.basename(sourcePath, ".sol") ||
+            artifact.contractName === path.basename(sourcePath, ".sol")
+          ) {
```

Comment:
Could you explain why it's okay to be looser with this check?

Reply:
you can't really compare the old check with the new check. the function works differently than it did before, so you really should be evaluating this PR for the current contents and not as a diff

i believe i used the `path.basename` here because `sourcePath` in this context varied. one can include `.sol` whereas the other does not

however it's been 5 months since i wrote this code and i'm not entirely sure what the driving force would be for this. i don't think this check is looser, but the correct amount of flexibility given the new execution context

by don't take my word for it and debug through the code to understand what happens. i'm not trying to sign up for maintaining this code, but just trying to put some work into performance since no one was jumping at the oppotunity

Reply:
> i believe i used the `path.basename` here because `sourcePath` in this context varied. one can include `.sol` whereas the other does not

I verified that this is correct. `importedSourcePath` (formally known as `sourcePath`) is the string in the `artifacts.require()` causing it to not be exactly  like `Artifact.sourcePath`

The function changed substantially causing the old school of thought of `sourcePath` being the source disk location to no longer be valid

Reply:
Is this the part of the logic that handles the canonical case of `artifacts.require("<contract-name>")`, or is that also covered by your initial short-circuit check at the beginning?

I think it's probably worth making a separate short-circuit check at the very beginning for the basic contract name case, since that's the "correct way", so it should also be the fastest. Also would be good for clarity around the various cases.

Reply:
> Is this the part of the logic that handles the canonical case of artifacts.require("<contract-name>"), or is that also covered by your initial short-circuit check at the beginning?

The latter.

The short-circuit covers:
- The canonical `artifacts.require("ContractName")`
- The less used `artifacts.require("ContractName.sol")`. In this case, we ignore the `.sol` the user provided and interpret it as `"ContractName"`
- The totally incorrect usage: `artifacts.require("some/relative/path/to/ContractName.sol")`. In this case, we ignore the path and `.sol` extension and interpret it as `"ContractName"`

> I think it's probably worth making a separate short-circuit check at the very beginning for the basic contract name case, since that's the "correct way", so it should also be the fastest. Also would be good for clarity around the various cases.

If I'm understanding you correctly, this suggestion is already implemented given my above answer